Objects/perf_trampoline.c

```
@@ -0,0 +1,233 @@
+        memcpy(memory + i * code_size, start, code_size * sizeof(char));
+    }
+
+    mprotect(memory, mem_size, PROT_READ | PROT_EXEC);
+
+    code_arena_t *new_arena = PyMem_RawCalloc(1, sizeof(code_arena_t));
+    if (new_arena == NULL) {
+        Py_FatalError("Failed to allocate new code arena struct");
+        return -1;
+    }
+
+    new_arena->start_addr = memory;
+    new_arena->current_addr = memory;
+    new_arena->size = mem_size;
+    new_arena->size_left = mem_size;
+    new_arena->code_size = code_size;
+    new_arena->prev = code_arena;
+    code_arena = new_arena;
+    return 0;
+}
+
+static void
+free_code_arenas(void)
+{
+    code_arena_t *cur = code_arena;
+    code_arena_t *prev;
+    code_arena = NULL; // invalid static pointer
+    while(cur) {
+        munmap(cur->start_addr, cur->size);
+        prev = cur->prev;
+        PyMem_RawFree(cur);
+        cur = prev;
+    }
+}
+
+static inline py_trampoline
+code_arena_new_code(code_arena_t *code_arena)
+{
+    py_trampoline trampoline = (py_trampoline)code_arena->current_addr;
+    code_arena->size_left -= code_arena->code_size;
+    code_arena->current_addr += code_arena->code_size;
+    return trampoline;
+}
+
+static inline py_trampoline
+compile_trampoline(void)
+{
+    if ((code_arena == NULL) || (code_arena->size_left <= code_arena->code_size)) {
+        if (new_code_arena() < 0) {
+            return NULL;
+        }
+    }
+
+    assert(code_arena->size_left <= code_arena->size);
+    return code_arena_new_code(code_arena);
+}
+
+static inline FILE *
+perf_map_get_file(void)
+{
+    if (perf_map_file) {
+        return perf_map_file;
+    }
+    char filename[100];
+    pid_t pid = getpid();
+    // TODO: %d is incorrect if pid_t is long long
+    snprintf(filename, sizeof(filename), "/tmp/perf-%d.map", pid);
+    perf_map_file = fopen(filename, "a");
+    if (!perf_map_file) {
+        _Py_FatalErrorFormat(__func__, "Couldn't open %s: errno(%d)", filename, errno);
```

If map file opening calls fail it is probably best to just disable the profiling rather than reporting the error and crashing. That isn't so important for the basic -Xperf dedicated profiled process case, but for turning it on and off at runtime you never want turning it on to crash the program. A `-v` verbose warning with error details instead of the fatal error would work to help people debug if they're surprised about no map existing.

---

It also avoids DoS scenarios where something just creates a bunch of /tmp/perf-XXX.map files and blocks the start of any python process with perf profiling enabled.

---

We can maybe throw an unraisable exception instead. I still think we should give more visibility to this error even when -v is not used.